core/src/main/java/hudson/slaves/SlaveComputer.java

```diff
@@ -482,7 +481,15 @@ public void onClosed(Channel c, IOException cause) {
                     taskListener.getLogger().println("Connection terminated");
                 }
                 closeChannel();
-                launcher.afterDisconnect(SlaveComputer.this, taskListener);
+                try {
+                    launcher.afterDisconnect(SlaveComputer.this, taskListener);
+                } catch (Throwable t) {
+                    LogRecord lr = new LogRecord(Level.SEVERE,
+                            "Launcher {0}'s afterDisconnect method propagated an exception when {1}'s connection was closed: {2}");
+                    lr.setThrown(t);
+                    lr.setParameters(new Object[]{launcher, SlaveComputer.this, t.getMessage()});
```

Comment:
Plain old

```java
logger.log(Level.SEVERE, "Launcher " + launcher + "'s afterDisconnect method propagated an exception when " + SlaveComputer.this + "'s connection was closed", t)
```

would suffice. The Jenkins `LogManager` offers no facility for suppressing stack traces so there is no reason to worry about it. And delaying string concatenation is a pointless optimization when `SEVERE` is logged by default.